### ISSUE TITLE
populate command needs unit-testing

### ISSUE THREAD

**Issue by yarikoptic** [dandi/dandisets#139]
continuation from https://github.com/dandi/dandisets/pull/137#issuecomment-1070943139

**Comment by jwodder** [comment]
Will not start until after #134 is merged.

**Comment by yarikoptic** [comment]
#134 was merged

**Comment by jwodder** [comment]
@yarikoptic Doing this will require making the remote details configurable.  I'm planning on moving these details — along with the Dandi instance, S3 bucket name, and GitHub organization names — to a YAML config file which will be stored in the repo.  However, I'm not sure what to do about configuring the local directories in which to store Dandiset & Zarr backups.  Currently, these are specified via a global `--target` option plus a non-global `--zarr-target` option that is only used where needed.  This seems needlessly complicated, and I'm considering moving both paths to the config file, but I'm not sure if it's a good idea to have such deployment-specific paths checked into Git.

Another possibility is to just have a single global command-line option specifying the parent directory of both the Dandisets folder and the Zarrs folder, which will then always be named "dandisets" and "dandizarrs", respectively.

Thoughts?  Other ideas?

**Comment by yarikoptic** [comment]
Since eventually similar setup could be adopted by someone else, concentrating configuration (for remote and local aspects) into a config file sounds like a great idea to me!
What I typically dislike about large "monolithic" configs though is that for some operation which needs to know only about a few settings, it would require instantiation of the rest of the infrastructure. So if e.g. `zarr_target` config setting is not needed for `populate` it better not be enforced to be present.  But otherwise -- I am totally ok with moving them all into a config file.  CLI options should remain there to assist with troubleshooting (e.g. running `update-from-backup` only on some specific dandiset).
As long as there is no secrets in configs, I think it is totally fine to save that configuration in our `dandisets` superdataset, even though it might contain local paths which would be specific to drogon.

Joining `target` and `zarr-target` into pointing to parent directory -- I have no strong opinion. IMHO it is ok to enforce such organization but it would be odd to point "above and beyond" to the folder which could potentially (like in our case) contain lots of other stuff.

**Comment by jwodder** [comment]
@yarikoptic 

> CLI options should remain there to assist with troubleshooting (e.g. running `update-from-backup` only on some specific dandiset).

The ability to specify Dandisets on the command line isn't going anywhere.  Exactly which options do you need to remain as CLI options?

**Comment by yarikoptic** [comment]
it is largely up to you since you would be seeing the picture better. But I see such options as `--s3bucket`, `--target` , `--zarr-target`, `--dandi-instance`, `backup_remote`, `--zarr-backup-remote`, `--zarr-gh-org`, `--gh-org`  to go into config; and such as `--log-level`, `--jobs`, `[ZARRS]`, `[DANDISETS]` `--exclude` -- remain CLI options.

**Comment by jwodder** [comment]
@yarikoptic How exactly do I set up a Git special remote for testing `populate`?  I tried following [these instructions](https://git-annex.branchable.com/special_remotes/git/) by doing:

```
git init --bare remote
( cd remote; git annex init )

git init repo
cd repo
git annex init
echo 'This is test text.' > file.txt
git commit -m Init file.txt

git remote add remote-git /path/to/remote
git annex initremote testremote type=git location=/path/to/remote autoenable=true
```

but the final command failed with "git-annex: could not find existing git remote with specified location".

**Comment by yarikoptic** [comment]
- better use `datalad.api.create` to create test datasets since that is what we have/use
- where do we have `type=git` special remote ATM in the workflow to bother setting it up?

```shell
(dandisets) dandi@drogon:/mnt/backup/dandi/dandisets/000108$ git show git-annex:remote.log
727f466f-60c3-4778-90b2-b2332856c2f8 chunk=1GB embedcreds=no encryption=none externaltype=rclone name=dandi-dandisets-dropbox prefix=dandi-dandisets/annexstore rclone_layout=lower target=dandi-dandisets-dropbox type=external timestamp=1622561489.11987s
(dandisets) dandi@drogon:/mnt/backup/dandi/dandisets/000108$ git remote
dandi-dandisets-dropbox
github
```
- seems we need some "backup" git-annex remote, can be `type=directory` and then a regular git (but git-ignore'ed) in place of `github` one (or I am ok even to test against real interactions with github in some dedicated for that organization)

**Comment by jwodder** [comment]
@yarikoptic I did:

```
mkdir remote
datalad create repo
cd repo
echo 'This is test text.' > file.txt
git commit -m Init file.txt
git annex initremote testremote type=directory directory=/path/to/remote encryption=none
```

and that worked, but then `git annex move --to testremote` did nothing, presumably because no files were added to the annex.  However, adding a PNG and commiting it — either with `git commit` or `datalad save` — also failed to add any files to the annex.  What exactly am I doing wrong?

**Comment by yarikoptic** [comment]
re your original question about git special remote -- I am as confused as you are!  don't have an answer

<details>
<summary>here is a turnkey reproducer</summary> 

```bash
#!/bin/bash

cd "$(mktemp -d ${TMPDIR:-/tmp}/dl-XXXXXXX)"
set -eux

git init --bare remote
( cd remote; git annex init )
rpath=$PWD/remote

git init repo
cd repo
git annex init
echo 'This is test text.' > file.txt
git add file.txt
git commit -m Init file.txt

git remote add remote-git $rpath
git annex initremote testremote type=git location=$rpath autoenable=true
```
</details>

may be @joeyh has a clue ?

**Comment by yarikoptic** [comment]
use `datalad save` or don't forget `git annex add` before `git commit`

**Comment by jwodder** [comment]
@yarikoptic It appears that `datalad save` doesn't add the file to the annex if it was added with `git add` instead of `git annex add`.  Is that supposed to happen?

**Comment by yarikoptic** [comment]
I think so since with `git add` you are instructing it to be added into `git`... I think there was a moment in time when git-annex would have added to `annex` but unlocked but then people complained and it was reverted... or I could be wrong

re `initremote` on type=git, looking at elderly version of code https://hackage.haskell.org/package/git-annex-5.20140926/src/Remote/Git.hs comment
```
 - For initremote, the git remote must already be set up, and have a uuid.
```
gives answer -- needed to have uuid already!!!! ... I think worth bug report to joey on a number of issue... preparing

**Comment by yarikoptic** [comment]
filed https://git-annex.branchable.com/bugs/initremote_type__61__git_is_not_working_for_unkn_reason/?updated

edit: related issue in datalad core land -- https://github.com/datalad/datalad/issues/6199
